libvast/src/system/index.cpp

```diff
@@ -38,172 +44,145 @@
-      VAST_ERROR(self,
-                 "failed to load statistics:", self->system().render(err));
-      return err;
-    }
-    VAST_DEBUG(self, "loaded statistics");
-  }
-  if (auto fname = meta_index_filename(); exists(fname)) {
-    VAST_VERBOSE(self, "loads meta index from", fname);
+  if (auto fname = index_filename(); exists(fname)) {
+    VAST_VERBOSE(self, "loads state from", fname);
     auto buffer = io::read(fname);
     if (!buffer) {
-      VAST_ERROR(self, "failed to read meta index file:",
+      VAST_ERROR(self, "failed to read index file:",
                  self->system().render(buffer.error()));
       return buffer.error();
     }
-    auto bytes = span<const byte>{*buffer};
-    if (auto err = fbs::unwrap<fbs::MetaIndex>(bytes, meta_idx))
-      return err;
-    VAST_DEBUG(self, "loaded meta index");
-  }
-  return caf::none;
-}
-
-caf::error index_state::flush_meta_index() {
-  VAST_VERBOSE(self, "writes meta index to", meta_index_filename());
-  auto flatbuf = fbs::wrap(meta_idx, fbs::file_identifier);
-  if (!flatbuf)
-    return flatbuf.error();
-  return io::save(meta_index_filename(), as_bytes(*flatbuf));
-}
-
-caf::error index_state::flush_statistics() {
-  VAST_VERBOSE(self, "writes statistics to", statistics_filename());
-  return save(&self->system(), statistics_filename(), stats);
-}
-
-caf::error index_state::flush_to_disk() {
-  VAST_TRACE("");
-  auto flush_all = [this]() -> caf::error {
-    // Flush meta index to disk.
-    if (auto err = flush_meta_index())
-      return err;
-    // Flush statistics to disk.
-    if (auto err = flush_statistics())
-      return err;
-    // Flush active partition.
-    if (active != nullptr)
-      if (auto err = active->flush_to_disk())
-        return err;
-    // Flush all unpersisted partitions. This only writes the meta state of
-    // each partition. For actually writing the contents of each INDEXER we
-    // need to rely on messaging.
-    for (auto& kvp : unpersisted)
-      if (auto err = kvp.first->flush_to_disk())
-        return err;
-    return caf::none;
-  };
-  if (auto err = flush_all()) {
-    VAST_ERROR(self, "failed to flush state:", self->system().render(err));
-    return err;
+    // TODO: Create a `index_ondisk_state` struct and move this part of the
+    // code into an `unpack()` function.
+    auto index = fbs::GetIndex(buffer->data());
+    auto version = index->version();
+    if (version != fbs::Version::Index_v0)
+      return make_error(ec::format_error,
+                        "unsupported index version, either remove the existing "
+                        "vast.db "
+                        "directory or try again with a newer version of VAST");
+    auto meta_idx = index->meta_index();
+    VAST_ASSERT(meta_idx);
+    auto error = unpack(*meta_idx, this->meta_idx);
+    if (error)
+      return error;
+    auto partition_uuids = index->partitions();
+    VAST_ASSERT(partition_uuids);
```

Comment:
is this also valid with 0 partitions?

Reply:
Yes. It was remarkably hard to figure this out from documentation, but I modified one of the unit tests to produce an empty vector and the pointer was still valid.